### PR TITLE
@Html.Raw(null) throws exception

### LINKED ISSUE #4350: @Html.Raw(null) stops rendering the view
I have a view which outputs raw HTML for a text editor:

	@Html.Raw(Model.Description)

Description is often null. With RC1 this worked. With the nightly built from last week the output which is received by the browser stops some lines (some 100 bytes) above the code with @Html.Raw(null). No exception. Just missing HTML output. The output stops even within HTML tags or attributes.

Workaround:

	@Html.Raw(Model.Description ?? "")

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Fixes #4350.

Also made sure MvcSandbox includes its views as content.

### CODE REVIEW

**Comment by ryanbrandenburg** [comment]
CC @rynowak since it looks like he wrote this class.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedStringWriterTest.cs [hunk 1]
```diff
@@ -44,6 +44,24 @@ static PagedBufferedStringWriterTest()
...
+
+            var writer = new PagedBufferedTextWriter(pool, inner);
+
+            // Act
+            writer.Write((char[])null);
+
+            await writer.FlushAsync();
+
+            // Assert
+            Assert.True(string.IsNullOrEmpty(inner.ToString()));
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedStringWriterTest.cs]: Stop and think before you use `Assert.True` or `Assert.False` - it's almost always wrong.

Why is it null or empty?

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedStringWriterTest.cs [hunk 2]
```diff
@@ -44,6 +44,24 @@ static PagedBufferedStringWriterTest()
         }
 
         [Fact]
+        public async Task Write_CharArray_Null()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedStringWriterTest.cs]: What about the other overload?

**Comment by rynowak** [comment]
:watch:

**Comment by ryanbrandenburg** [comment]
:up: :date:

File: samples/MvcSandbox/project.json [hunk 3]
```diff
@@ -17,6 +17,9 @@
     "Microsoft.Extensions.Configuration.Json": "1.0.0-*",
     "Microsoft.Extensions.Logging.Console": "1.0.0-*"
   },
+  "content": [
+    "Views"
+  ],
```
Comment (dougbu) [inline samples/MvcSandbox/project.json]: Why change the sandbox?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs [hunk 4]
```diff
@@ -83,11 +83,21 @@ public override void Write(char value)
...
+
             Write(buffer, 0, buffer.Length);
         }
 
         public override void Write(char[] buffer, int index, int count)
         {
+            if (buffer == null)
+            {
+                return;
+            }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: This method should match `TextWriter` and / or `StringWriter`. Both of those implementations of this method will throw an `ArgumentNullException` if `buffer == null`. Do the same in this block.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedTextWriterTest.cs [hunk 5]
```diff
@@ -83,6 +101,23 @@ static PagedBufferedStringWriterTest()
...
+        public async Task Write_CharArray_Bounded_Null()
+        {
+            // Arrange
+            var pool = new TestArrayPool();
+            var inner = new StringWriter();
+
+            var writer = new PagedBufferedTextWriter(pool, inner);
+
+            // Act
+            writer.Write(null, 0, 0);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedTextWriterTest.cs]: `Assert.Throws<ArgumentNullException>(() => writer.Write(null, 0, 0), ...)`

**Comment by dougbu** [comment]
Curious: Where was the NRE in the #4350 scenario caught? I'm a bit surprised _any_ content reached the browser.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs (hunk 4 above)
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: Ah, you mean just this overload should throw. Seems reasonable since that's what the other writers do. I'll also take the other methods null checking structure. Functionally it'll be the same but it can't hurt to standardize.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedStringWriterTest.cs (hunk 2 above)
Comment (ryanbrandenburg) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedStringWriterTest.cs]: Done.

File: samples/MvcSandbox/project.json (hunk 3 above)
Comment (ryanbrandenburg) [inline samples/MvcSandbox/project.json]: I was using the Sandbox to test the bug and my fix in the real world and figured reverting the stuff I did to get it to work in the new world was a waste.

**Comment by ryanbrandenburg** [comment]
As for where the NRE was caught, it looks like everything above the offending function was written but nothing after.

**Comment by dougbu** [comment]
> it looks like everything above the offending function was written but nothing after.

You've described the behaviour. I was asking why. For example, is there a `try` / `catch` / ignore (in MVC) on the stack, does the view buffer flush to the response when it's disposed, ...?

File: samples/MvcSandbox/project.json (hunk 3 above)
Comment (dougbu) [inline samples/MvcSandbox/project.json]: :ok:

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs [hunk 6]
```diff
@@ -110,25 +118,28 @@ public override void Write(char[] buffer, int index, int count)
 
         public override void Write(string value)
         {
-            var index = 0;
-            var count = value.Length;
-
-            while (count > 0)
+            if (value != null)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: Was more readable before: Exit early to avoid the extra nesting.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs [hunk 7]
```diff
@@ -83,11 +83,19 @@ public override void Write(char value)
...
+            {
+                Write(buffer, 0, buffer.Length);
+            }
         }
 
         public override void Write(char[] buffer, int index, int count)
         {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer), Resources.ArgumentNull_Buffer);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: Get rid of the resource and just throw `ArgumentNullException(nameof(buffer))`. We fortunately don't use this constructor overload anywhere in MVC.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx [hunk 8]
```diff
@@ -280,4 +280,7 @@
   <data name="ValueInterfaceAbstractOrOpenGenericTypesCannotBeActivated" xml:space="preserve">
     <value>The type '{0}' cannot be activated by '{1}' because it is either a value type, an interface, an abstract class or an open generic type.</value>
   </data>
+  <data name="ArgumentNull_Buffer" xml:space="preserve">
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx]: Revert changes to this file.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs [hunk 9]
```diff
@@ -83,129 +83,146 @@ public override void Write(char value)
...
         public override void Write(char[] buffer)
         {
+            if (buffer == null)
+            {
+                return;
+            }
+
             Write(buffer, 0, buffer.Length);
         }
+    }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: Huh?

File: samples/MvcSandbox/Startup.cs [hunk 10]
```diff
@@ -5,6 +5,7 @@
 using Microsoft.AspNetCore.Hosting;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
+using System.IO;
```
Comment (dougbu) [inline samples/MvcSandbox/Startup.cs]: `System` `using`s first.

**Comment by ryanbrandenburg** [comment]
@dougbu after stepping through this and consulting with @rynowak I've determined that the reason it behaves this way is that the exception is being thrown AFTER the response has already begun. Razor has constructed it's list of write operations, but not executed them and we've already sent all the headers (including 200 OK) to the client. Now we get to the list of operations from Razor which it thinks are safe, which includes our Write.Raw(null) and it throws an exception. Having already sent the response code we can't very well change it to a 500 now, so Kestrel cuts it's losses and we get an "ERR_INCOMPLETE_CHUNKED_ENCODING" and Chrome trying to show us the parts of the page it DID get.

So basically it behaves this way instead of giving us an error because the exception is thrown in a low enough level function that there's no good way to throw an error remaining. If this exception had been thrown by user code in the cshtml page rather than inside an actual Write we would have gotten the 500 and white page that I believe you expected.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs [hunk 11]
```diff
@@ -83,129 +83,146 @@ public override void Write(char value)
...
-                    page,
-                    _currentIndex,
-                    copyLength);
-
-                _currentIndex += copyLength;
-                index += copyLength;
-
-                count -= copyLength;
-            }
+            throw new ArgumentNullException(nameof(buffer));
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: @rynowak could you to confirm that having this throw is the correct behavior? It seems pretty dangerous because it causes the response to fail after we've already sent the headers but @dougbu is correct that that's what we do in `System.IO.TextWriter`.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: Argument validation is argument validation. If we have a condition that trips this, let's throw earlier and protect ourselves

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs [hunk 12]
```diff
@@ -83,11 +83,21 @@ public override void Write(char value)
...
+            }
+
             Write(buffer, 0, buffer.Length);
         }
 
         public override void Write(char[] buffer, int index, int count)
         {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: @rynowak could you to confirm that having this throw is the correct behavior? It seems pretty dangerous because it causes the response to fail after we've already sent the headers but @dougbu is correct that that's what we do in System.IO.TextWriter.

P.S.: Sorry for the dupe, I accidentally staled the other comment to this effect.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: Odd that this comment moved to another overload. But thanks for correcting the `Write(char[], int, int)` method.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs (hunk 12 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs]: @ryanbrandenburg your changes make this class act like almost every other `TextWriter`. That fixes inconsistencies as well as the `Raw()` problem. Having this method do anything other than `throw` for `buffer == null` creates inconsistencies and likely doesn't affect any scenario except user code calling this method directly. Even then, it just makes the behaviour of that code vary depending on exactly where in the `.cshtml` file it appears.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedTextWriterTest.cs [hunk 13]
```diff
@@ -83,6 +102,19 @@ static PagedBufferedStringWriterTest()
...
+        public void Write_CharArray_Bounded_Null()
+        {
+            // Arrange
+            var pool = new TestArrayPool();
+            var inner = new StringWriter();
+
+            var writer = new PagedBufferedTextWriter(pool, inner);
+
+            // Act & Assert
+            Assert.Throws<ArgumentNullException>(() => writer.Write(null, 0, 0));
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PagedBufferedTextWriterTest.cs]: Pass the `paramName` argument i.e. `Assert.Throws<ArgumentNullException>("buffer", () => writer.Write(null, 0, 0));`.

**Comment by dougbu** [comment]
:shipit: with one last test fix.
